### ISSUE TITLE
Support for NASA's Common Data Format (CDF)?

### ISSUE THREAD

**Issue by mathewjowens** [sunpy/sunpy#2285]
Are there any plans to provide support fro NASA's [Common Data Format (CDF)](https://cdf.gsfc.nasa.gov/)? It's the standard -- and often only -- format for [all NASA in-situ data](https://cdaweb.sci.gsfc.nasa.gov/index.html/).

There was a [python CDF library](https://pypi.python.org/pypi/CDF), but it's no longer supported and refuses to compile on most machines. Similarly, LANL's spacepy library subsumed this CDF library, but it itself is also no longer supported and very difficult to get up and running.

**Comment by Cadair** [comment]
We have been talking about this in the PlasmaPy chat recently. I had a quick look at the CDF library in SpacePy, it does not look like it would be fun to extract the CDF support.

The best solution in my opinion is to fix the packaging issue for SpacePy, I think @SolarDrew was planning on taking a shot at getting SpacePy in conda.

**Comment by mathewjowens** [comment]
David Stansby (Imperial), who's started work on [HelioPy](https://pypi.python.org/pypi/heliopy). has put together a version of SpacePy's CDF library with somewhat reduced dependencies. But ultimately, it's the same underlying code and, yeah, best avoided.

After a bit of trial and error, I managed to get SpacePy to install under ubuntu, but it wasn't happy with any combination of Windows compilers. So I suspect there's some nasty OS-specific stuff to be dealt with. SpacePy in conda seems like a good solution. 

Something of a hollow offer, given my low level of expertise, but let me know if there's anything I can help with!

**Comment by Cadair** [comment]
Oh no, not windows being a pain. That is never fun for anyone.

Do you know if/how dead SpacePy is? It would be a shame if it was completely left to die. The actual CDF bindings code in SpacePy is pretty clever (not pretty, but clever) I would not want to attempt to write my own from scratch.

**Comment by mathewjowens** [comment]
(I managed to get SpacePy to compile under the Windows 10 linux subsystem. Two, poorly meshed OSs. What fun.)

SpacePy - not sure. I might shoot Jon Niehof and Steve Morley an email, as I know them both. The main LANL SpacePy page is down, which isn't encouraging, but according to github, there's been activity reasonably recently.

**Comment by mathewjowens** [comment]
I got in contact with Steve Morley (smorley@lanl.gov) and Jon Niehof (Jonathan.Niehof@unh.edu). They're happy to chat about CDF support, etc. But SpacePy is still being supported and developed (the loss of webpage is for other reasons). In particular, they're already looking at getting the Windows support fixed up.

**Comment by SolarDrew** [comment]
I was indeed planning on conda-ing up SpacePy. It might take a while - I don't have much experience with conda packaging and CDF looks like it'll be a whole load of hassle - but I'll keep you in the loop.

**Comment by mathewjowens** [comment]
Excellent, thank you!

**Comment by mathewjowens** [comment]
FYI: Thanks to David Stansby's help, I've got HelioPy installed and hence python reading CDFs under Windows. If you're looking to incorporate CDF under SunPy or PlasmaPy, it would be well worth speaking to David.

**Comment by dstansby** [comment]
Hello, David here! (I have stumbled upon this issue). I have started writing https://github.com/heliopython/heliopy - the idea behind it was sort of "SunPy for in-situ data".

re. CDF files, I have extracted `pycdf` from `SpacePy` and put it in it's own module in `heliopy`. I have yet to work out exactly what it depends on, but installing and using it seems to work fine for me (and I think @mathewjowens now). It is certainly much easier to install than `SpacePy`

It also sounds like I should start chatting to `sunpy` and `plasmapy` people to make sure
- I'm not duplicating any work in those packages
- Those packages aren't duplicating any of my work

I'll try and reach out by email this afternoon to get a conversation started. If anyone wants to contact me re. `pycdf` or `heliopy` feel free at dstansby@gmai.com.
